contracts/EtomicSwapNft.sol

```diff
@@ -0,0 +1,719 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.20;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
+import "@openzeppelin/contracts/token/ERC721/IERC721Receiver.sol";
+import "@openzeppelin/contracts/token/ERC1155/IERC1155.sol";
+import "@openzeppelin/contracts/token/ERC1155/IERC1155Receiver.sol";
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+
+contract EtomicSwapNft is ERC165, IERC1155Receiver, IERC721Receiver {
+    using SafeERC20 for IERC20;
+
+    enum MakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerSpent,
+        MakerRefunded
+    }
+
+    struct MakerPayment {
+        bytes20 paymentHash;
+        uint32 paymentLockTime;
+        MakerPaymentState state;
+    }
+
+    event MakerPaymentSent(bytes32 id);
+    event MakerPaymentSpent(bytes32 id);
+    event MakerPaymentRefundedTimelock(bytes32 id);
+    event MakerPaymentRefundedSecret(bytes32 id);
+
+    mapping(bytes32 => MakerPayment) public makerPayments;
+
+    enum TakerPaymentState {
+        Uninitialized,
+        PaymentSent,
+        TakerApproved,
+        MakerSpent,
+        TakerRefunded
+    }
+
+    struct TakerPayment {
+        bytes20 paymentHash;
+        uint32 preApproveLockTime;
+        uint32 paymentLockTime;
+        TakerPaymentState state;
+    }
+
+    event TakerPaymentSent(bytes32 id);
+    event TakerPaymentApproved(bytes32 id);
+    event TakerPaymentSpent(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedSecret(bytes32 id, bytes32 secret);
+    event TakerPaymentRefundedTimelock(bytes32 id);
+
+    mapping(bytes32 => TakerPayment) public takerPayments;
+
+    address public immutable dexFeeAddress;
+
+    constructor(address feeAddress) {
+        require(
+            feeAddress != address(0),
+            "feeAddress must not be zero address"
+        );
+
+        dexFeeAddress = feeAddress;
+    }
+
+    function spendErc721MakerPayment(
```

Comment:
We should follow the same order as `EtomicSwapV2.sol` where we start with the function that recieves the maker paymet https://github.com/KomodoPlatform/etomic-swap/blob/124445a6f1cd61892bdcaa3cec786eee91b7e5f5/contracts/EtomicSwapV2.sol#L61
This order follows the swap steps order which makes it easier to read the code in order.

Reply:
We use `"safeTransferFrom"` methods from ERC721 and ERC1155 ABI to `send payment`